File: 2019/day24.py
import itertools
import sys
from pprint import pprint

inp="""\
...#.
#.##.
#..##
#.###
##..."""

test = """\
....#
#..#.
#..##
..#..
#...."""

def makegrid(txt):
    grid = []
    for ln in txt.splitlines():
        grid.append( list(".#".index(c) for c in ln) )
    return grid


def rating(grid):
    return sum( cell << i for i,cell in enumerate(sum(grid,[])))

valid = (0,1,2,3,4)

checks = ((-1,0),(1,0),(0,-1),(0,1))

def cycle(grid):
    new = []
    for y in range(len(grid)):
        row = []
        for x in range(len(grid[0])):
            neighbors = sum(
                (grid[y+dy][x+dx] 
                for dx,dy in checks
                if x+dx in valid and y+dy in valid)
            )
            if neighbors == 1 or (neighbors == 2 and grid[y][x] == 0):
                row.append(1)
            else:
                row.append(0)
        new.append( row )
    return new


if 'test' in sys.argv:
    grid = makegrid(test)
else:
    grid = makegrid(inp)
pprint( grid )
print( rating( grid) )

values = {rating(grid): 0}
for i in itertools.count(1):
    grid = cycle(grid)
    pprint( grid )
    r = rating(grid)
    print( i, r )
    if rating(grid) in values:
        print( r )
        break
    values[r] = i

print( "Part 1:", i, rating(grid) )

File: 2019/day24.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <algorithm>
#include <numeric>

const char * input = 
    "...#."
    "#.##."
    "#..##"
    "#.###"
    "##...";

const char * test = 
    "....#"
    "#..#."
    "#..##"
    "..#.."
    "#....";

bool TRACE = false;
bool TEST = false;

class Grid
{
    std::vector<uint8_t> m_Grid;
    std::vector<std::pair<int,int>> m_Checks;

    uint8_t at(int x, int y )
    {
        if( (x >= 0) && (x < 5) && (y >= 0) && (y < 5) )
            return m_Grid[y*5+x];
        else
            return 0;
    }

public:
    Grid( const std::string txt )
    {
        m_Grid.resize( txt.size() );
        std::transform( txt.begin(), txt.end(), m_Grid.begin(),
            [](char c) { return c == '#'; } );

//        for( ; *txt; txt++ )
//            m_Grid.push_back( *txt == '#' );

        m_Checks.emplace_back( -1, 0 );
        m_Checks.emplace_back( 1, 0 );
        m_Checks.emplace_back( 0, -1 );
        m_Checks.emplace_back( 0, 1 );
    }

    int rating()
    {
        int i = 0;
        return std::accumulate( m_Grid.begin(), m_Grid.end(), 0,
            [&i](int sum, uint8_t next) { return sum + (next << i++); } );
    }

    void cycle()
    {
        std::vector<uint8_t> grid;
        for( int y = 0; y < 5; y++ )
        {
            for( int x = 0; x < 5; x++ )
            {
                int neighbors = std::accumulate( 
                    m_Checks.begin(), m_Checks.end(), 0,
                    [this,x,y](int i, std::pair<int,int> c)
                    { return i+at(x+c.first,y+c.second); }
                );

                grid.push_back(
                    neighbors == 1 ||  (neighbors == 2 && at(x,y) == 0) 
                );
            }
        }
        m_Grid = grid;
    }

    void print()
    {
        for( int y = 0; y < 5; y++ )
        {
            for( int x = 0; x < 5; x++ )
                std::cout << " " << ".#"[at(x,y)];
            std::cout <<"\n";
        }
    }
};


void main( int argc, char ** argv )
{
    while( *++argv )
    {
        if( strcmp( *argv, "trace" ) == 0 )
            TRACE = true;
        else if( strcmp( *argv, "test" ) == 0 )
            TEST = true;
    }

    Grid grid( TEST ? test : input );
    grid.print();
    int r = grid.rating();
    std::cout << grid.rating() << "\n";
    std::map<int, int> track;
    track[r] = 0;
    for( int i = 1; ; i++ )
    {
        grid.cycle();
        grid.print();
        r = grid.rating();
        std::cout << i << " " << r << "\n";
        if( track.find(r) != track.end() )
        {
            std::cout << "Part 1: " << i << " " << r << "\n";
            break;
        }
        track[r] = i;
    }
}
